File: algorithms/scala/BinarySearch.scala

/**
 * Perform a binary search, in this case find 10 in [5, 10, 15, 20, 25].
 */
object BinarySearch extends Application {

  println( binarySearch( Array[Int](5, 10, 15, 20, 25), 10 ))

  /**
   * Perform a binary search for an Int.
   *
   * @param array the array to search.
   * @param term the term to search for.
   * @return true if found, false if not.
   */
  def binarySearch( array: Array[Int], term: Int ): Boolean = {

    var low: Int = 0
    var hi: Int = array.length - 1

    while( low <= hi ) {

      val mid: Int = ( low + hi ) / 2

      if ( term > array( mid ) ) {

        low = mid + 1

      } else if ( term < array( mid ) ) {

        hi = mid - 1

      } else {

        return true

      }

    }

    false

  }

}

File: algorithms/ruby/binary_search.rb
module BinarySearch

  public

  # Perform a binary search for an Int.
  #
  def b_contains?(arr, term)
    return false if arr == nil || arr.length < 1

    len = arr.length
    lo = 0
    hi = len - 1

    while hi > lo do
      mid = (lo + hi) / 2

      return true if arr[mid] == term

      if term > arr[mid]
        lo = mid + 1
      elsif term < arr[mid]
        hi = mid - 1
      end
    end

    false
  end

  private

end

if __FILE__ == $0

  include BinarySearch

  # Perform a binary search

  puts b_contains? [5, 10, 15, 20, 25, 30, 35], 10
  puts b_contains? [5, 10, 15, 20, 25, 30, 35], 11

end